src/renderer/atlas/AtlasEngine.cpp

```diff
@@ -0,0 +1,1449 @@
+                    /* baseFontCollection */ fontCollection.get(),
+                    /* baseFamilyName     */ _api.fontMetrics.fontName.get(),
+                    /* baseWeight         */ baseWeight,
+                    /* baseStyle          */ baseStyle,
+                    /* baseStretch        */ DWRITE_FONT_STRETCH_NORMAL,
+                    /* mappedLength       */ &mappedLength,
+                    /* mappedFont         */ font.addressof(),
+                    /* scale              */ &scale));
+
+                mappedFontFace.reset();
+                if (font)
+                {
+                    THROW_IF_FAILED(font->CreateFontFace(mappedFontFace.addressof()));
+                }
+            }
+
+            mappedEnd = idx + mappedLength;
+
+            if (!mappedFontFace)
+            {
+                auto pos1 = idx;
+                auto col1 = _api.bufferLineColumn[pos1];
+                for (auto pos2 = idx + 1; pos2 <= mappedEnd; ++pos2)
+                {
+                    if (const auto col2 = _api.bufferLineColumn[pos2]; col1 != col2)
+                    {
+                        static constexpr auto replacement = L'\uFFFD';
+                        _emplaceGlyph(nullptr, scale, pos1, pos2);
+                        pos1 = pos2;
+                        col1 = col2;
+                    }
+                }
+
+                continue;
+            }
+        }
+        else
+        {
+            if (!mappedFontFace)
+            {
+                const auto baseWeight = _api.attributes.bold ? DWRITE_FONT_WEIGHT_BOLD : static_cast<DWRITE_FONT_WEIGHT>(_api.fontMetrics.fontWeight);
+                const auto baseStyle = _api.attributes.italic ? DWRITE_FONT_STYLE_ITALIC : DWRITE_FONT_STYLE_NORMAL;
+
+                wil::com_ptr<IDWriteFontFamily> fontFamily;
+                THROW_IF_FAILED(fontCollection->GetFontFamily(0, fontFamily.addressof()));
+
+                wil::com_ptr<IDWriteFont> font;
+                THROW_IF_FAILED(fontFamily->GetFirstMatchingFont(baseWeight, DWRITE_FONT_STRETCH_NORMAL, baseStyle, font.addressof()));
+
+                THROW_IF_FAILED(font->CreateFontFace(mappedFontFace.put()));
+            }
+
+            mappedEnd = gsl::narrow_cast<u32>(_api.bufferLine.size());
+        }
+
+        // We can reuse idx here, as it'll be reset to "idx = mappedEnd" in the outer loop anyways.
+        for (u32 complexityLength = 0; idx < mappedEnd; idx += complexityLength)
+        {
+            BOOL isTextSimple;
+            THROW_IF_FAILED(_sr.textAnalyzer->GetTextComplexity(_api.bufferLine.data() + idx, mappedEnd - idx, mappedFontFace.get(), &isTextSimple, &complexityLength, _api.glyphIndices.data()));
+
+            if (isTextSimple)
+            {
+                for (size_t i = 0; i < complexityLength; ++i)
+                {
+                    _emplaceGlyph(mappedFontFace.get(), scale, idx + i, idx + i + 1u);
+                }
+            }

```

Comment:
I feel like this doesn't save us that much in the majority case because of the regional variations in i/j that ruin it for Cascadia Code, the default font.

Reply:
Cascadia Mono is the default font isn't it? That one doesn't have the i/j issue and all basic ASCII characters are considered "simple" for that font (which makes it a lot faster than Cascadia Code). 🙂